[Feature] Support list partition in be & add/drop list partition

Astralidea opened the pull request:
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #4035

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

Smith-Cruise committed:
[Refactor] Refactor partial selected for complex typ

Signed-off-by: Astralidea <astralidea@163.com>

amber-create committed:
Update BrokerLoad.md (#16711)

Signed-off-by: Astralidea <astralidea@163.com>

Astralidea committed:
Support list partition in be

Signed-off-by: Astralidea <astralidea@163.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 7 / 15 (46.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|15|46.67%|[769, 770, 771, 772, 773, 774, 775, 777]|

wangruin committed:
[Feature] Add docker compose and more dockerfiles (#16670)

Add docker compose and more dockerfiles

Signed-off-by: Astralidea <astralidea@163.com>

Astralidea committed:
support add list partition

Signed-off-by: Astralidea <astralidea@163.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 8 / 20 (40.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|0|4|00.00%|[3058, 3059, 3060, 3061]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|15|46.67%|[769, 770, 771, 772, 773, 774, 775, 777]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|

Astralidea committed:
support drop list partition

Signed-off-by: Astralidea <astralidea@163.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 22 / 39 (56.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|0|4|00.00%|[3058, 3059, 3060, 3061]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|21|34|61.76%|[769, 770, 771, 772, 773, 774, 775, 777, 857, 858, 860, 861, 865]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|

satanson committed:
fixup one-phase aggregation fails to use query cache (#16705)

Signed-off-by: Astralidea <astralidea@163.com>

kevincai committed:
[UT] fix build break (#16732)

Signed-off-by: Astralidea <astralidea@163.com>

amber-create committed:
remove conditional updates for broker load (#16731)

Signed-off-by: Astralidea <astralidea@163.com>

kevincai committed:
[BugFix] Use member comparison to implement operator== (#16733)

Signed-off-by: Astralidea <astralidea@163.com>

GavinMar committed:
[BugFix] Fix the thrift library conflict in some cases by linking the required cachelib dependencies statically (#16674)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Fix] Choose non-polymorphyic function first (#16701)

Signed-off-by: Astralidea <astralidea@163.com>

motto1314 committed:
[Doc]Add datadog-connector&superset-connector doc (#16727)

Signed-off-by: Astralidea <astralidea@163.com>

luohaha committed:
[Feature] lake pk table compaction (#16212)

Signed-off-by: Astralidea <astralidea@163.com>

abc982627271 committed:
[BugFix] throw NPE when add a new partition to a lake table, then restart fe and drop this partition (#16706)

Signed-off-by: Astralidea <astralidea@163.com>

ZiheLiu committed:
[Enhancement] Show default_wg in audit log (#16743)

Signed-off-by: Astralidea <astralidea@163.com>

Smith-Cruise committed:
[BugFix] Avoid lambda query failed (#16749)

Signed-off-by: Astralidea <astralidea@163.com>

wyb commented on be/src/exec/tablet_info.cpp:
```diff
@@ -315,17 +336,18 @@ Status OlapTablePartitionParam::find_tablets(Chunk* chunk, std::vector<OlapTable
         ChunkRow row;
         row.columns = &partition_columns;
         row.index = 0;
+        bool is_by_value_partition = _t_param.partitions[0].__isset.in_keys;
```
is_list_partition?
as a member variable?

wyb commented on fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:
```diff
@@ -826,24 +836,34 @@ private void dropPartition(long dbId, String partitionName, boolean isForceDrop,
+            } else if (partitionInfo.getType() == PartitionType.LIST) {
+                ListPartitionInfo listPartitionInfo = (ListPartitionInfo) partitionInfo;
+                if (!isForceDrop) {
+                    // TODO: recycle list partition
```
report an error?

amber-create committed:
fix typo in broker load docs (#16758)

Signed-off-by: Astralidea <astralidea@163.com>

chaoyli committed:
[Refactor] Add a configure to enable and disable auto tablet distribution(#16702)

1. If the user does not specify the tablet number upon creating a table or adding a partition, we will auto-determine the tablet distribution.
2. If the table exists, the user wants to use the feature. Change the enable_auto_tablet_distribution to true.
3. If the user wants to disable the auto distribution, specify the tablet num manually after setting enable_auto_tablet_distribution to false.

Signed-off-by: Astralidea <astralidea@163.com>

ABingHuang committed:
[BugFix] add MVActiveChecker for external table (#16746)

* add MVActiveChecker for external table. external tables' metas are not persistent, so the check should be done every time when cache is refresh. now it is done by async, which will be optimized later.

Signed-off-by: Astralidea <astralidea@163.com>

liuyehcf committed:
[Enhancement] profile merge mechanism enhancement (#16717)

Signed-off-by: Astralidea <astralidea@163.com>

evelynzhaojie committed:
2.5 release notes and other bugs (#16782)

Signed-off-by: Astralidea <astralidea@163.com>

Youngwb committed:
[Enhancement] Support struct type partial selected with aggregate function (#16779)

Signed-off-by: Astralidea <astralidea@163.com>

satanson committed:
[Bugfix] column names must be associated with slotIds when normalizing OlapScanNode accessing unpartitioned table (#16754)

Signed-off-by: Astralidea <astralidea@163.com>

nshangyiming committed:
[Feature] set GlobalStateMgr.USING_NEW_PRIVILEGE to true (#16556)

Fix all the failed UTs.

After this pr, all the codes which need to check privilege for
table or db or mv etc. should use new privilege framework.

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
[Doc] fix document with Chinese word (#16790)

Signed-off-by: Astralidea <astralidea@163.com>

chaoyli committed:
Update str_to_date.md (#16789)

Signed-off-by: Astralidea <astralidea@163.com>

sduzh committed:
[UT] Remove legacy code (#16794)

Signed-off-by: Astralidea <astralidea@163.com>

Linkerist committed:
[UT] Add some memtable nullable sort key case. (#15683)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Refactor] Remove unused use_vecotorized (#16793)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
Update JAVA_UDF.md (#16798)

Signed-off-by: Astralidea <astralidea@163.com>

chaoyli committed:
[Refator] Fix the wrong log in updateVersionInfoForRecovery() (#16617)

Signed-off-by: Astralidea <astralidea@163.com>

stdpain committed:
[BugFix] Fix projection may reuse column ref when prune union empty (#16646)

Signed-off-by: Astralidea <astralidea@163.com>

Smith-Cruise committed:
[Doc] Add doc about IDE setup (#16810)

Signed-off-by: Astralidea <astralidea@163.com>

satanson committed:
[BugFix] correct errors of NOT(EQ_FOR_NULL) expr (#16805)

Signed-off-by: Astralidea <astralidea@163.com>

zuyu committed:
[BugFix] ChunkHelper: convert_schema_to_format_v2 -> convert_schema (#16812)

Signed-off-by: Astralidea <astralidea@163.com>

sevev committed:
[BugFix] PersistentIndex may save index data that has been deleted when allowing early flush (#16774)

Signed-off-by: Astralidea <astralidea@163.com>

gengjun-git committed:
[BugFix]Fix truncate table throws NPE bug (#16775)

Signed-off-by: Astralidea <astralidea@163.com>

zuyu committed:
[BugFix] Fixed tscancode errors (#16816)

Signed-off-by: Astralidea <astralidea@163.com>

zuyu committed:
[Refactor] Using Filter for duplicated Column::Filter (#16815)

Signed-off-by: Astralidea <astralidea@163.com>

xiaoyong-z committed:
[Refactor] update md5sum for broker_thirdparty_jars.tar.gz (#16826)

Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Doc] Release 2.4.3 (#16828)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Refactor] Change docker compose image to version 2.5 (#16830)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Refactor] Use default conf in docker-file (#16831)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Docs] Add document for deploying starrocks in docker (#16832)

Signed-off-by: Astralidea <astralidea@163.com>

xiaoyong-z committed:
[Enhancement] optimize memcpy in bytebuffer (#16772)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Refactor] Make AWS SDK library path independent (#16837)

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
fix document ident error (#16838)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Feature] Add dockerfiles to build dev docker image on ubuntu (#16845)

Signed-off-by: Astralidea <astralidea@163.com>

titianqx committed:
[Doc] fix quick start part in toc.md (#16847)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Feature] Add dev-toolchain-ubuntu docker (#16853)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Refactor] Refactor build-dev (#16855)

Signed-off-by: Astralidea <astralidea@163.com>

imay committed:
[Refactor] Use lld as default linker (#16856)

Co-authored-by: Alvin <alvin.c.zhao@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

zuyu committed:
[Refactor] Use be_http_port, instead of webserver_port (#16792)

For consistency, use be_http_port, instead of be_web_port

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Feature] Add be fe docker files (#16858)

Signed-off-by: Astralidea <astralidea@163.com>

alvin-coding committed:
[Refactor] Change compose to main (#16859)

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
Use system timezone instead of default timezone (#16844)

Signed-off-by: Astralidea <astralidea@163.com>

chaoyli committed:
[BugFix] Fix the version fall back when the RPC lag (#16841)

If the replica version in the FE catches up to 10, it should reject the smaller version(less than 10) reported by RPC.
The PRC may lag due to the high traffic network jam.  If the report is not rejected, the version falling back results in the commit error like the following.
```
 Commit failed. txn: 1002 table: test1 tablet: 10016 quorum: 0<1 errorReplicas: 10017:{be:10002 172.26.92.139 V:9 LFV:10},]}}, msg={replay end load job}
```

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
[Enhancement] fix logging of cancel MV (#16842)

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
[BugFix] fix NLJ stuck for empty output (#16875)

Signed-off-by: Astralidea <astralidea@163.com>

chaoyli committed:
[Refactor] Rename VectorizedField to Field and VectorizedSchema to Schema (#16862)

Rename VectorizedField to Field, unify the two nouns.
Rename VectorizedSchema to Schema, unify the two nouns.

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
[BugFix] validate array syntax in casting string to array (#16866)

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
[Doc] improve document of resource group (#16880)

Signed-off-by: Astralidea <astralidea@163.com>

murphyatwork committed:
[BugFix] Fix convert json object with empty key (#16852)

Signed-off-by: Astralidea <astralidea@163.com>

Youngwb committed:
[Enhancement] Support CTE for Trino parser on starrocks (#16877)

Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Tool] dev-env docker image with third-party dependencies and maven dependencies (#16890)

Signed-off-by: Deng Liu <d.liu@celonis.de>

## Which issues of this PR fixes ：
- Get all the  dependencies pre-built/downloaded into the dev-env image to speed up the build of fe and be
- Image has been pushed to my personal repo on dockerhub and verified it can successfully build sr fe/be with significant speed improvement.
https://hub.docker.com/r/dengliu/sr-dev-env/tags

## Problem Summary(Required) ：
Get all the  dependencies pre-built/downloaded into the dev-env image to speed up the build of fe and be

Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Doc] Add broker config (#16906)

Signed-off-by: EsoragotoSpirit <wanglichen@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

evelynzhaojie committed:
[Doc] Update rn and create table (#16914)

* Update release-2.5.md

* Update CREATE TABLE.md

* Update hive_catalog.md

Signed-off-by: Astralidea <astralidea@163.com>

amber-create committed:
[Doc] update aws s3 parameter names in broker load docs (#16927)

Signed-off-by: amber-create <yangyanping@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

silverbullet233 committed:
[BugFix] Disable RewriteMinMaxAggToMetaScanRule when table has deletion (#16898)

Signed-off-by: silverbullet233 <3675229+silverbullet233@users.noreply.github.com>
Signed-off-by: Astralidea <astralidea@163.com>

ZiheLiu committed:
[BugFix] Init scan executor for compute node (#16933)

Signed-off-by: ZiheLiu <ziheliu1024@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

miomiocat committed:
[Feature] Support INT64 based timestamp for JNI connector (#16920)

Signed-off-by: miomiocat <284487410@qq.com>
Signed-off-by: Astralidea <astralidea@163.com>

kangkaisen committed:
[BugFix] Fix bitnot largeint cast (#16928)

Signed-off-by: kangkaisen <kangkaisen@apache.org>
Signed-off-by: Astralidea <astralidea@163.com>

ZiheLiu committed:
[BugFix] Set name of MetaScanOperator (#16936)

Signed-off-by: ZiheLiu <ziheliu1024@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

decster committed:
[BugFix] Fix error in incremental clone introduced by #15935 (#16930)

PR #15935 introduced a new mechanism that an incremental clone can fall back to a full clone if there are too many rowsets need to be copied. But the fallback code return a NotFound error rather than OK, this will cause clone to fail.

```
I0127 00:46:02.271173  7878 snapshot_manager.cpp:112] make primary snapshot tablet:334728 cur_version:65905 missing_version_ranges:65662 timeout:180
I0127 00:46:02.271206  7878 tablet_updates.cpp:2954] get_rowsets_for_snapshot: too many rowsets for incremental clone #rowset:244 #rowset_for_full_clone:1 tablet:334728 #version:1 [65905.1 65905.1@0 65905.1] #pending:0
W0127 00:46:02.271214  7878 agent_server.cpp:308] fail to make_snapshot. tablet_id:334728 msg:Not found: get_rowsets_for_snapshot: too many rowsets for incremental clone #rowset:244 #rowset_for_full_clone:1 tablet:334728 #version:1 [65905.1 65905.1@0 65905.1] #pending:0
```
This PR fixes this bug.

Signed-off-by: Binglin Chang <decstery@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

hellolilyliuyi committed:
change-toc-of-create-routine-load (#16943)

Signed-off-by: Astralidea <astralidea@163.com>

xiaoyong-z committed:
[Enhancement] Upgrade broker's aliyun-sdk from 2.7.2 to jindosdk-4.6.2 (#16935)

Signed-off-by: Astralidea <astralidea@163.com>

sduzh committed:
[BugFix] Schema change job cannot be cancelled (#16787)

Fix a lake table's schema change job of state FINISHED_REWRITTING cannot
be cancelled even after the table has been dropped.

Signed-off-by: sduzh <zhuming9011@qq.com>
Signed-off-by: Astralidea <astralidea@163.com>

amber-create committed:
[Doc] modify s3 param names (#16960)

Signed-off-by: amber-create <yangyanping@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

liuyehcf committed:
[Enhancement] Replace BitSet with RoaringBitMap to lower memory usage (#16499)

Signed-off-by: Astralidea <astralidea@163.com>

GavinMar committed:
[BugFix] Avoid calling `GetReadStatistics` function when accessing juicefs (#16931)

Signed-off-by: GavinMar <yangguansuo@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Doc] Fix Lint errors (#16965)

Signed-off-by: EsoragotoSpirit <wanglichen@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

xiaoyong-z committed:
[Enhancement] support broker load from azure abfs (#16760)

In this pr, we support broker load from azure abfs.
We remove the check of path prefix in getFileSystem, if the path prefix is not hdfs/s3a/oss/cos...,
we will try to create a generic file system. The reason why we can do this is because hadoop/s3 SDK
is compatible with all file/object storage system.
Then if in future, if there is a request for supporting new file system,
we only need to add the settings in hdfs-site.xml, core-site.xml of BE/FE, and the jar lib in the specific place.

Signed-off-by: xyz <a997647204@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

evelynzhaojie committed:
[Doc] trim and create table (#16982)

* [Doc] trim and create table

* Signed-off-by:
evelynzhaojie
<zhaojie@starrocks.com>

* Signed-off-by:
evelynzhaojie
<zhaojie@starrocks.com>

Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Doc] Add details for BACKUP and MV (#16973)

Signed-off-by: EsoragotoSpirit <wanglichen@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Doc] Fix MV desc and link (#16974)

Signed-off-by: EsoragotoSpirit <wanglichen@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Doc] Fix issues in Config and RG (#16983)

Signed-off-by: EsoragotoSpirit <wanglichen@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

nshangyiming committed:
[Feature] Fix submit task privilege error (#16907)

PrivilegeCheckerV2 should also support unqualified table name.

Signed-off-by: Astralidea <astralidea@163.com>

stdpain committed:
[Enhancement] Reduce memory usage in full sorter (#16937)

1. avoid copy and reserve extra chunk when merge sorted chunks
2. release memory in SortRun::reset

Signed-off-by: stdpain <drfeng08@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

ABingHuang committed:
fix unstable refresh materialized view ut

Signed-off-by: Astralidea <astralidea@163.com>

amber-create committed:
[Doc] update EXPORT (#16992)

Signed-off-by: amber-create <yangyanping@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

luohaha committed:
[BugFix] fix be crash when schema change (#16999)

When concurrent run data loading and schema change, BE may crash at:
```
 0x57baa51 starrocks::vectorized::SchemaChangeHandler::_convert_historical_rowsets()
*** SIGSEGV (@0x40) received by PID 13522 (TID 0x7fc2bb3a0700) from PID 64; stack trace: ***
    @          0x6a6b792 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7fc3fd261630 (unknown)
    @          0x57baa51 starrocks::vectorized::SchemaChangeHandler::_convert_historical_rowsets()
    @          0x57bd560 starrocks::vectorized::SchemaChangeHandler::_do_process_alter_tablet_v2_normal()
    @          0x57be888 starrocks::vectorized::SchemaChangeHandler::_do_process_alter_tablet_v2()
    @          0x57bed8a starrocks::vectorized::SchemaChangeHandler::process_alter_tablet_v2()
    @          0x55dde96 starrocks::EngineAlterTabletTask::execute()
    @          0x53faeee starrocks::StorageEngine::execute_task()
    @          0x3a4171e starrocks::run_alter_tablet_task()
    @          0x5b4c405 starrocks::ThreadPool::dispatch_thread()
    @          0x5b4712a starrocks::Thread::supervise_thread()
    @     0x7fc3fd259ea5 start_thread
    @     0x7fc3f293f9fd __clone
    @                0x0 (unknown)
```
Because in _convert_historical_rowsets, assume that sc_params.rowset_readers and sc_params.rowsets_to_change have same valid items. If don't, crash will happen.

So I add checking code when geneate rowsets_to_change, when base_tablet->get_rowset_by_version can't find rowset by version, print all helpful message to check what happen and just return error to prevent BE from crash. Still need find out in which situation, version from capture_consistent_versions by max version can't find rowset by get_rowset_by_version.

Signed-off-by: luohaha <18810541851@163.com>
Signed-off-by: Astralidea <astralidea@163.com>

ZiheLiu committed:
[BugFix] Own short_key_encode result for LogicalSplitMorselQueue (#16977)

Signed-off-by: ZiheLiu <ziheliu1024@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

Vision0220 committed:
Update README.md (#16979)

Signed-off-by: Astralidea <astralidea@163.com>

Youngwb committed:
[UT] Fix unstable ut (#17003)

Signed-off-by: Youngwb <yangwenbo_mailbox@163.com>
Signed-off-by: Astralidea <astralidea@163.com>

LiShuMing committed:
Remove redundent withMaterializedStatementView/withNewMaterializedView methods in StarRocksAssert (#16985)

Signed-off-by: Kevin Li <ming.moriarty@gmail.com>
Signed-off-by: Astralidea <astralidea@163.com>

Commit:
[Doc] Add Upgrade Info (#16988)

Signed-off-by: EsoragotoSpirit <wanglichen@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

nshangyiming committed:
[Enhancement] Refine the error msg of drop func and other priv checking (#16980)

Signed-off-by: Dejun Xia <xiadejun@starrocks.com>
Signed-off-by: Astralidea <astralidea@163.com>

silverbullet233 committed:
[Enhancement] add MemUsageBytes when showing current queries (#16923)

Signed-off-by: silverbullet233 <3675229+silverbullet233@users.noreply.github.com>
Signed-off-by: Astralidea <astralidea@163.com>

nshangyiming committed:
[BugFix] Restore logging for cluster load statistics (#17000)

The logging for cluster load statistics is mistakenly removed by previous pr.

Signed-off-by: Astralidea <astralidea@163.com>

Astralidea committed:
Add the restriction that null is not allowed to be created.

Signed-off-by: Astralidea <astralidea@163.com>
